cdc/sink/producer/kafka/kafka.go

```diff
@@ -151,49 +153,28 @@ func (k *kafkaSaramaProducer) SyncBroadcastMessage(ctx context.Context, message
 }
 
 func (k *kafkaSaramaProducer) Flush(ctx context.Context) error {
-	targetOffsets := make([]uint64, k.partitionNum)
-	for i := 0; i < len(k.partitionOffset); i++ {
-		targetOffsets[i] = atomic.LoadUint64(&k.partitionOffset[i].sent)
-	}
+	down := make(chan struct{}, 1)
 
-	noEventsToFLush := true
-	for i, target := range targetOffsets {
-		if target > atomic.LoadUint64(&k.partitionOffset[i].flushed) {
-			noEventsToFLush = false
-			break
-		}
-	}
-	if noEventsToFLush {
-		// no events to flush
-		return nil
+	k.mu.Lock()
+	inflight := k.mu.inflight
+	immediateFlush := inflight == 0
+	if !immediateFlush {
+		k.mu.flushDown = down
 	}
+	k.mu.Unlock()
 
-	// checkAllPartitionFlushed checks whether data in each partition is flushed
-	checkAllPartitionFlushed := func() bool {
-		for i, target := range targetOffsets {
-			if target > atomic.LoadUint64(&k.partitionOffset[i].flushed) {
-				return false
-			}
-		}
-		return true
+	if immediateFlush {
```

Comment:
A race condition may happen in this place. If `k.mu` is unlocked and immediately locked by another goroutine doing `AsyncSendMessage`,  the `k.mu.inflight` will be incremented. As a result, the value hold by `immediateFlush` is no longer correct.